synapse/handlers/presence.py

```diff
@@ -249,6 +257,29 @@ async def process_replication_rows(self, token, rows):
         """Process presence stream rows received over replication."""
         pass
 
+    async def maybe_send_presence_to_interested_destinations(
+        self, states: List[UserPresenceState]
+    ):
+        """If this instance is a federation sender, send the states to all
+        destinations that are interested.
+        """
+
+        if not self._send_federation:
+            return
```

Comment:
still think you should move this out to the callers and rename the function `send_presence_to_interested_destinations`; then you can use it from `PresenceHandler._update_states`. I might be missing something though, or you might already have plans here, or something.

Reply:
There are a couple of things that make me a bit hesitant to do that: 1) in the end state the federation ping code won't hit the federation sender, but instead the federation queue, and so everywhere that calls this function will end up having to do that check. 2) We don't always have a federation sender, so we need to do *some* check, whether its an assert or a no-op (and bear in mind that this will soon turn back into a `if self._federation` check again).

I think it'll be easier to discuss this at the next PR, as it'll be clearer (for both of us) what the end result feels like